monai/networks/nets/milmodel.py

```diff
@@ -0,0 +1,206 @@
+import os
+from enum import Enum
+
+import torch
+import torch.nn as nn
+import torchvision.models as models
+
+
+class MilMode(str, Enum):
+    MEAN = "mean"
+    MAX = "max"
+    ATT = "att"
+    ATT_TRANS = "att_trans"
+    ATT_TRANS_PYRAMID = "att_trans_pyramid"
+
+    def __str__(self) -> str:
+        return str.__str__(self)
+
+
+class MILModel(nn.Module):
+    """
+    A wrapper around backbone classification model suitable for MIL
+
+    Args:
+        num_classes: number of output classes
+        mil_mode: MIL variant (supported max, mean, att, att_trans, att_trans_pyramid
+        pretrained: init backbone with pretrained weights. Defaults to True.
+        backbone: Backbone classifier CNN. Defaults to None, it which case ResNet50 will be used.
+        backbone_nfeatures: Number of output featues of the backbone CNN (necessary only when using custom backbone)
+
+    mil_mode:
+        mean - average features from all instances, equivalent to pure CNN (non MIL)
+        max - retain only the instance with the max probability for loss calculation
+        att - attention based MIL https://arxiv.org/abs/1802.04712
+        att_trans - transformer MIL
+        att_trans_pyramid - transformer pyramid MIL
+
+    """
+
+    def __init__(
+        self,
+        num_classes: int,
+        mil_mode: MilMode = MilMode.ATT,
+        pretrained: bool = True,
+        backbone: nn.Module = None,
+        backbone_nfeatures: int = None,
+        trans_blocks: int = 4,
+        trans_dropout: float = 0.0,
+    ) -> None:
+
+        super().__init__()
+
+        self.mil_mode = mil_mode
+        print("MILModel with mode", mil_mode, "num_classes", num_classes)
+        self.attention = None
+
+        if backbone is None:
+            # use ResNet50 backbone
+
+            # os.environ["TORCH_HOME"] = "../../torchhome"
+            # import torchvision.models as models
+
+            net = models.resnet50(pretrained=pretrained)
+            nfc = net.fc.in_features  # save the number of final features
+            net.fc = torch.nn.Identity()  # remove final linear layer
+
+            self.extra_outputs = {}
+            if self.mil_mode == MilMode.ATT_TRANS_PYRAMID:
```

Comment:
Are you going to have ` MilMode.ATT_TRANS_PYRAMID` only when the backbone is None?

Reply:
yes, that mode we can only support for backbone None,  it requires knowledge of the model structure and it's difficult to support it for an arbitrary model. In fact, I'm not sure if we need to support arbitrary model at all

Reply:
If `MILModel` is not going to work with arbitrary models (which makes sense), can you specify with which models it can work? 
Thanks

Reply:
Currently, it works with any backbone (if someone removes FC layer manually, and provides number of output features),
The exception will be  MilMode.ATT_TRANS_PYRAMID which doesn't support arbitrary models at all 

The alternative is we don't support Arbitrary models at all